dev/benchmarks/macrobenchmarks/lib/src/web/recorder.dart

```diff
@@ -332,175 +406,168 @@ class _WidgetBuildRecorderHostState extends State<_WidgetBuildRecorderHost> {
   }
 }
 
-/// Pumps frames and records frame metrics.
-abstract class Recorder {
-  Recorder._(this.name);
+/// Series of time recordings indexed in time order.
+///
+/// It can calculate [average], [standardDeviation] and [noise]. If the amount
+/// of data collected is higher than [_kMeasuredSampleCount], then these
+/// calculations will only apply to the latest [_kMeasuredSampleCount] data
+/// points.
+class Timeseries {
+  Timeseries();
 
-  /// The name of the benchmark being recorded.
-  final String name;
+  final List<num> _values = <num>[];
 
-  /// Frame metrics recorded during a single benchmark run.
-  final List<FrameMetrics> _frames = <FrameMetrics>[];
+  /// The total amount of data collected, including ones that were dropped
+  /// because of the sample size limit.
+  int count = 0;
 
-  /// Runs the benchmark and records a profile containing frame metrics.
-  Future<Profile> run();
+  double get average => _computeMean(_values);
 
-  /// Decides whether the data collected so far is sufficient to stop, or
-  /// whether the benchmark should continue collecting more data.
-  ///
-  /// The signals used are sample size, noise, and duration.
-  bool _shouldContinue() {
-    // Run through a minimum number of frames.
-    if (_frames.length < _kMinSampleCount) {
-      return true;
-    }
+  double get standardDeviation =>
+      _computeStandardDeviationForPopulation(_values);
 
-    final Profile profile = _generateProfile();
-
-    // Is it still too noisy?
-    if (profile.drawFrameDurationNoise > _kNoiseThreshold) {
-      // If the benchmark has run long enough, stop it, even if it's noisy under
-      // the assumption that this benchmark is always noisy and there's nothing
-      // we can do about it.
-      if (_frames.length > _kMaxSampleCount) {
-        print(
-          'WARNING: Benchmark noise did not converge below ${_kNoiseThreshold * 100}%. '
-          'Stopping because it reached the maximum number of samples $_kMaxSampleCount. '
-          'Noise level is ${profile.drawFrameDurationNoise * 100}%.',
-        );
-        return false;
-      }
+  double get noise => standardDeviation / average;
 
-      // Keep running.
-      return true;
+  void add(num value) {
+    count++;
+    _values.add(value);
+    // Don't let the list grow beyond [_kMeasuredSampleCount].
+    if (_values.length > _kMeasuredSampleCount) {
+      _values.removeAt(0);
```

Comment:
It would be useful for debugging to have all the data (e.g. to answer the question "How did my benchmark's noise converge?"). Maybe have two fields `allValues` and `measuredValues`?